src/db/controller/channel.ts

```diff
@@ -0,0 +1,312 @@
+  }
+
+  public static delOpenEvents(events: ChannelOpenCloseTable[]) {
+    if (events.filter((event) => event.id === undefined).length !== 0) {
+      throw new Error('id must be exists to remove channel on open')
+    }
+
+    if (events.length === 0) return
+
+    del<ChannelOpenCloseTable>(
+      DB,
+      this.tableName,
+      events.map((v) => ({ id: v.id as number }))
+    )
+  }
+
+  public static updateInProgress(id?: number, inProgress = true) {
+    update<ChannelOpenCloseTable>(
+      DB,
+      this.tableName,
+      { in_progress: inProgress ? Boolean.TRUE : Boolean.FALSE },
+      [{ id }]
+    )
+  }
+
+  public static resetPacketInProgress(db?: Database) {
+    db = db ?? DB
+    update<ChannelOpenCloseTable>(db, this.tableName, {
+      in_progress: Boolean.FALSE,
+    })
+  }
+
+  private static async feedChannelOpenInitEvent(
+    lcd: LCDClient,
+    chainId: string,
+    event: ChannelOpenCloseEvent
+  ): Promise<() => void> {
+    const connection = await ConnectionController.getConnection(
+      lcd,
+      chainId,
+      event.channelOpenCloseInfo.srcConnectionId
+    )
+
+    // add channel on open for dst chain
+    const channelOnOpen: ChannelOpenCloseTable = {
+      in_progress: Boolean.FALSE,
+      height: event.channelOpenCloseInfo.height,
+      state: ChannelState.INIT,
+      chain_id: connection.counterparty_chain_id,
+      connection_id: connection.counterparty_connection_id,
+      port_id: event.channelOpenCloseInfo.dstPortId,
+      channel_id: event.channelOpenCloseInfo.dstChannelId,
+      counterparty_chain_id: chainId,
+      counterparty_connection_id: event.channelOpenCloseInfo.srcConnectionId,
+      counterparty_port_id: event.channelOpenCloseInfo.srcPortId,
+      counterparty_channel_id: event.channelOpenCloseInfo.srcChannelId,
+    }
+
+    return () => {
+      insert(DB, this.tableName, channelOnOpen)
+    }
+  }
+
+  private static async feedChannelOpenTryEvent(
+    lcd: LCDClient,
+    chainId: string,
+    event: ChannelOpenCloseEvent
+  ): Promise<() => void> {
+    const connection = await ConnectionController.getConnection(
+      lcd,
+      chainId,
+      event.channelOpenCloseInfo.dstConnectionId
+    )
+
+    // add channel on open for src chain
+    const channelOnOpen: ChannelOpenCloseTable = {
+      in_progress: Boolean.FALSE,
+      height: event.channelOpenCloseInfo.height,
+      state: ChannelState.TRYOPEN,
+      chain_id: connection.counterparty_chain_id,
+      connection_id: connection.counterparty_connection_id,
+      port_id: event.channelOpenCloseInfo.srcPortId,
+      channel_id: event.channelOpenCloseInfo.srcChannelId,
+      counterparty_chain_id: chainId,
+      counterparty_connection_id: event.channelOpenCloseInfo.dstConnectionId,
+      counterparty_port_id: event.channelOpenCloseInfo.dstPortId,
+      counterparty_channel_id: event.channelOpenCloseInfo.dstChannelId,
+    }
+
+    return () => {
+      del<ChannelOpenCloseTable>(DB, this.tableName, [
```

Comment:
no case for deleting wrong entry? don't we need to check `chain_id`?

Reply:
(`state`, `counterparty_chain_id`, `counterparty_channel_id`) is unique

Reply:
oh reasonable